## Ансамбли

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from matplotlib import pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [ ]:
!wget 'https://drive.google.com/uc?id=1UyGD0iuIsO0Q7mBn8YY6lSuCc8r5MGzH' -O train.csv

--2024-08-26 09:26:42--  https://drive.google.com/uc?id=1UyGD0iuIsO0Q7mBn8YY6lSuCc8r5MGzH
Resolving drive.google.com (drive.google.com)... 74.125.69.139, 74.125.69.101, 74.125.69.113, ...
Connecting to drive.google.com (drive.google.com)|74.125.69.139|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1UyGD0iuIsO0Q7mBn8YY6lSuCc8r5MGzH [following]
--2024-08-26 09:26:42--  https://drive.usercontent.google.com/download?id=1UyGD0iuIsO0Q7mBn8YY6lSuCc8r5MGzH
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 173.194.194.132, 2607:f8b0:4001:c10::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|173.194.194.132|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 460676 (450K) [application/octet-stream]
Saving to: ‘train.csv’

train.csv           100%[===================>] 449.88K  --.-KB/s    in 0.004s  

2024-08-26 09:26:45 (101 MB/s) - ‘train

In [ ]:
data = pd.read_csv('train.csv')
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(data.drop(columns=['SalePrice', 'Id']),
                                                    data['SalePrice'],
                                                    test_size=0.3,
                                                    random_state=42)
X_train.shape, X_test.shape

((1022, 79), (438, 79))

In [ ]:
categorials = X_train.select_dtypes('object').columns

X_train[categorials] = X_train[categorials].fillna('nan')
X_test[categorials] = X_test[categorials].fillna('nan')

In [ ]:
integer = X_train.select_dtypes(['int', 'float']).columns

X_train[integer] = X_train[integer].fillna(-999)

X_test[integer] = X_test[integer].fillna(-999)

In [ ]:
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder(sparse=False, handle_unknown='ignore')
ohe.fit(X_train[categorials])

X_train[ohe.get_feature_names_out()] = ohe.transform(X_train[categorials])
X_test[ohe.get_feature_names_out()] = ohe.transform(X_test[categorials])

X_train = X_train.drop(categorials, axis=1)
X_test = X_test.drop(categorials, axis=1)

X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
135           20         80.0    10400            7            6       1970   
1452         180         35.0     3675            5            5       2005   
762           60         72.0     8640            7            5       2009   
932           20         84.0    11670            9            5       2006   
435           60         43.0    10667            7            6       1996   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
135           1970       288.0           0           0  ...             0.0   
1452          2005        80.0         547           0  ...             0.0   
762           2009         0.0          24           0  ...             0.0   
932           2006       302.0           0           0  ...             0.0   
435           1996         0.0         385         344  ...             1.0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
135            0.0           0.0          1.0                    0.0   
1452           0.0           0.0          1.0                    0.0   
762            0.0           0.0          0.0                    0.0   
932            0.0           0.0          1.0                    0.0   
435            0.0           0.0          0.0                    0.0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
135                     0.0                   0.0                   0.0   
1452                    0.0                   0.0                   0.0   
762                     0.0                   0.0                   0.0   
932                     0.0                   0.0                   0.0   
435                     0.0                   0.0                   0.0   

      SaleCondition_Normal  SaleCondition_Partial  
135                    1.0                    0.0  
1452                   1.0                    0.0  
762                    1.0                    0.0  
932                    1.0                    0.0  
435                    1.0                    0.0  

[5 rows x 298 columns]

In [ ]:
X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
135           20         80.0    10400            7            6       1970   
1452         180         35.0     3675            5            5       2005   
762           60         72.0     8640            7            5       2009   
932           20         84.0    11670            9            5       2006   
435           60         43.0    10667            7            6       1996   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
135           1970       288.0           0           0  ...             0.0   
1452          2005        80.0         547           0  ...             0.0   
762           2009         0.0          24           0  ...             0.0   
932           2006       302.0           0           0  ...             0.0   
435           1996         0.0         385         344  ...             1.0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
135            0.0           0.0          1.0                    0.0   
1452           0.0           0.0          1.0                    0.0   
762            0.0           0.0          0.0                    0.0   
932            0.0           0.0          1.0                    0.0   
435            0.0           0.0          0.0                    0.0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
135                     0.0                   0.0                   0.0   
1452                    0.0                   0.0                   0.0   
762                     0.0                   0.0                   0.0   
932                     0.0                   0.0                   0.0   
435                     0.0                   0.0                   0.0   

      SaleCondition_Normal  SaleCondition_Partial  
135                    1.0                    0.0  
1452                   1.0                    0.0  
762                    1.0                    0.0  
932                    1.0                    0.0  
435                    1.0                    0.0  

[5 rows x 298 columns]

In [ ]:
X_test.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
892           20         70.0     8414            6            8       1963   
1105          60         98.0    12256            8            5       1994   
413           30         56.0     8960            5            6       1927   
522           50         50.0     5000            6            7       1947   
1036          20         89.0    12898            9            5       2007   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  ...  SaleType_ConLw  \
892           2003         0.0         663           0  ...             0.0   
1105          1995       362.0        1032           0  ...             0.0   
413           1950         0.0           0           0  ...             0.0   
522           1950         0.0         399           0  ...             0.0   
1036          2008        70.0        1022           0  ...             0.0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
892            0.0           0.0          1.0                    0.0   
1105           0.0           0.0          1.0                    0.0   
413            0.0           0.0          1.0                    0.0   
522            0.0           0.0          1.0                    0.0   
1036           0.0           0.0          1.0                    0.0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
892                     0.0                   0.0                   0.0   
1105                    0.0                   0.0                   0.0   
413                     0.0                   0.0                   0.0   
522                     0.0                   0.0                   0.0   
1036                    0.0                   0.0                   0.0   

      SaleCondition_Normal  SaleCondition_Partial  
892                    1.0                    0.0  
1105                   1.0                    0.0  
413                    1.0                    0.0  
522                    1.0                    0.0  
1036                   1.0                    0.0  

[5 rows x 298 columns]

### Voting

<img src='https://drive.google.com/uc?id=1se-9Q6cJfXh8mKeDNcQkfa5LfMT6OdYm' width=500>


In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.ensemble import VotingRegressor

vote = VotingRegressor(
    estimators=[
        ('dt', DecisionTreeRegressor()),
        ('lin', LinearRegression()),
        ('l2', Ridge())
    ]
)

In [ ]:
vote.fit(X_train, y_train)

VotingRegressor(estimators=[('dt', DecisionTreeRegressor()),
                            ('lin', LinearRegression()), ('l2', Ridge())])

In [ ]:
vote.score(X_train, y_train), vote.score(X_test, y_test)

(0.9717614317665708, 0.807834525970397)

In [ ]:
vote.named_estimators

{'dt': DecisionTreeRegressor(), 'lin': LinearRegression(), 'l2': Ridge()}

In [ ]:
for i in vote.named_estimators:
    print(f'train model {i} {vote.named_estimators_[i].score(X_train, y_train)}')
    print(f'test model {i} {vote.named_estimators_[i].score(X_test, y_test)}')

train model dt 1.0
test model dt 0.7762527460523947
train model lin 0.9405742053126412
test model lin 0.12454071580031201
train model l2 0.924130269961213
test model l2 0.8833066093272545


In [ ]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler


vote = VotingRegressor(
    estimators=[
        ('dt', DecisionTreeRegressor()),
        # ('lin', make_pipeline(StandardScaler(), LinearRegression())),
        ('l2', make_pipeline(StandardScaler(), Ridge()))
    ]
)

vote.fit(X_train, y_train)

VotingRegressor(estimators=[('dt', DecisionTreeRegressor()),
                            ('l2',
                             Pipeline(steps=[('standardscaler',
                                              StandardScaler()),
                                             ('ridge', Ridge())]))])

In [ ]:
vote.score(X_train, y_train), vote.score(X_test, y_test)

(0.9845420683393219, 0.8772518569519834)

In [ ]:
vote.named_estimators

{'dt': DecisionTreeRegressor(),
 'l2': Pipeline(steps=[('standardscaler', StandardScaler()), ('ridge', Ridge())])}

In [ ]:
for i in vote.named_estimators:
    print(f'train model {i} {vote.named_estimators_[i].score(X_train, y_train)}')
    print(f'test model {i} {vote.named_estimators_[i].score(X_test, y_test)}')

train model dt 1.0
test model dt 0.7715991789028007
train model l2 0.9381682733572878
test model l2 0.8649375634573391


In [ ]:
vote = VotingRegressor(
    estimators=[
        ('dt', DecisionTreeRegressor()),
        ('l2', make_pipeline(StandardScaler(), Ridge()))
    ],
    weights=[0.9, 1]
)

vote.fit(X_train, y_train)

VotingRegressor(estimators=[('dt', DecisionTreeRegressor()),
                            ('l2',
                             Pipeline(steps=[('standardscaler',
                                              StandardScaler()),
                                             ('ridge', Ridge())]))],
                weights=[0.9, 1])

In [ ]:
vote.score(X_train, y_train), vote.score(X_test, y_test)

(0.9828720978829052, 0.8794014144295303)

In [ ]:
for i in vote.named_estimators:
    print(f'train model {i} {vote.named_estimators_[i].score(X_train, y_train)}')
    print(f'test model {i} {vote.named_estimators_[i].score(X_test, y_test)}')

train model dt 1.0
test model dt 0.761093143390153
train model l2 0.9381682733572878
test model l2 0.8649375634573391


### Stacking


<img src='https://drive.google.com/uc?id=18ek8izAoN_M8YIvd6HOIL3ylHlcQIBJt' width=500>


In [ ]:
from sklearn.ensemble import StackingRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import RidgeCV


lr = LinearRegression()
stack = StackingRegressor(
    [
        ('LinearRegression', lr),
        ('KNeighborsRegressor', KNeighborsRegressor()),
        ('DecisionTree', DecisionTreeRegressor()),
    ], RidgeCV([1, 5, 10])
)

In [ ]:
stack.fit(X_train, y_train)

StackingRegressor(estimators=[('LinearRegression', LinearRegression()),
                              ('KNeighborsRegressor', KNeighborsRegressor()),
                              ('DecisionTree', DecisionTreeRegressor())],
                  final_estimator=RidgeCV(alphas=[1, 5, 10]))

In [ ]:
stack.final_estimator_.alpha_

10

In [ ]:
print(f'train {stack.score(X_train, y_train)}')
print(f'on test {stack.score(X_test, y_test)}')

train 0.9670062952149283
on test 0.7675776975117857


In [ ]:
for i in stack.named_estimators:
    print(f'train model {i} {stack.named_estimators_[i].score(X_train, y_train)}')
    print(f'test model {i} {stack.named_estimators_[i].score(X_test, y_test)}')

train model LinearRegression 0.9405742053126412
test model LinearRegression 0.12454071580031201
train model KNeighborsRegressor 0.7501837788800589
test model KNeighborsRegressor 0.676138613361793
train model DecisionTree 1.0
test model DecisionTree 0.7304804405860326


#### 🧠 Упражнение. Реализация своего стэкинга

1. Обучите базовые модели на тренировочной выборке и соберите новый датасет для мета-модели

In [ ]:
# Определение базовых моделей
model1 = LinearRegression()
model2 = DecisionTreeRegressor()
model3 = KNeighborsRegressor()

# Обучение базовых моделей
...

# Получение предсказаний базовых моделей на обучающей выборке
pred1_train = ...
pred2_train = ...
pred3_train = ...

In [ ]:
# Формирование нового обучающего набора для мета-модели
meta_X_train = ...
meta_X_train

array([[177007.38397421, 174000.        , 212680.        ],
       [125278.22135251, 145000.        , 132670.        ],
       [227898.64063741, 215200.        , 183418.        ],
       ...,
       [115196.97582276, 115000.        ,  97900.        ],
       [163433.74864486, 189950.        , 157090.        ],
       [189116.04125762, 174000.        , 180200.        ]])

2. Обучите мета-модель на новой тренировочной выборке

In [ ]:
# Обучение мета-модели
meta_model = RidgeCV([1, 5, 10])
...

RidgeCV(alphas=[1, 5, 10])

3. Получите предсказания не тестовой выборке базовыми моделями и сделайте итоговое предсказание мета-моделью

In [ ]:
# Получение предсказаний базовых моделей на тестовой выборке
pred1_test = ...
pred2_test = ...
pred3_test = ...

# Формирование нового тестового набора для мета-модели
meta_X_test = ...

# Прогнозирование с помощью мета-модели
y_pred = ...

# Оценка качества модели
r2_score(y_test, y_pred)

0.7453477941351987

##### 🧠 Упражнение (ответ). Реализация своего стэкинга

In [ ]:
# Определение базовых моделей
model1 = LinearRegression()
model2 = DecisionTreeRegressor()
model3 = KNeighborsRegressor()

# Обучение базовых моделей
model1.fit(X_train, y_train)
model2.fit(X_train, y_train)
model3.fit(X_train, y_train)

# Получение предсказаний базовых моделей на обучающей выборке
pred1_train = model1.predict(X_train)
pred2_train = model2.predict(X_train)
pred3_train = model3.predict(X_train)

In [ ]:
# Формирование нового обучающего набора для мета-модели
meta_X_train = np.column_stack((pred1_train, pred2_train, pred3_train))

# Обучение мета-модели
meta_model = RidgeCV([1, 5, 10])
meta_model.fit(meta_X_train, y_train)

RidgeCV(alphas=[1, 5, 10])

In [ ]:
# Получение предсказаний базовых моделей на тестовой выборке
pred1_test = model1.predict(X_test)
pred2_test = model2.predict(X_test)
pred3_test = model3.predict(X_test)

# Формирование нового тестового набора для мета-модели
meta_X_test = np.column_stack((pred1_test, pred2_test, pred3_test))

# Прогнозирование с помощью мета-модели
y_pred = meta_model.predict(meta_X_test)

# Оценка качества модели
r2_score(y_test, y_pred)

0.7676976993067

## Градиентный бустинг

[Видео](https://youtu.be/sDv4f4s2SB8) с подробным объяснением алгоритма

**Продемонстрируем работу бустинга**

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score


X, y = load_diabetes(return_X_y=True, as_frame=True)
X

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019907 -0.017646  
1   -0.039493 -0.068332 -0.092204  
2   -0.002592  0.002861 -0.025930  
3    0.034309  0.022688 -0.009362  
4   -0.002592 -0.031988 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018114  0.044485  
439 -0.011080 -0.046883  0.015491  
440  0.026560  0.044529 -0.025930  
441 -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

In [ ]:
y

0      151.0
1       75.0
2      141.0
3      206.0
4      135.0
       ...  
437    178.0
438    104.0
439    132.0
440    220.0
441     57.0
Name: target, Length: 442, dtype: float64

In [ ]:
# Обучаем первое дерево
tree1 = DecisionTreeRegressor(max_depth=3,
                              random_state=2)
tree1.fit(X, y)

prediction1 = tree1.predict(X)
print(f'R2 score {r2_score(y, prediction1)}')

R2 score 0.5006720154703376


In [ ]:
# Подсчитываем остатки
residual1 = y - prediction1
y[2], prediction1[2], residual1[2]

(141.0, 208.57142857142858, -67.57142857142858)

In [ ]:
# Обучаем второе дерево на ошибках предыдущих
tree2 = DecisionTreeRegressor(max_depth=3,
                              random_state=2)

tree2.fit(X, residual1)

prediction2 = tree1.predict(X) + tree2.predict(X)
print(f'R2 score {r2_score(y, prediction2)}')

R2 score 0.5785866108916171


In [ ]:
# Подсчитываем остатки
residual2 = y - prediction2
y[2], prediction2[2], residual2[2]

(141.0, 207.5052553799773, -66.50525537997731)

In [ ]:
# Обучаем третье дерево на ошибках предыдущих
tree3 = DecisionTreeRegressor(max_depth=3,
                              random_state=2)

tree3.fit(X, residual2)

prediction3 = tree1.predict(X) + tree2.predict(X) + tree3.predict(X)
print(f'R2 score {r2_score(y, prediction3)}')

R2 score 0.6271203539706915


In [ ]:
# Подсчитываем остатки
residual3 = y - prediction3
y[2], prediction3[2], residual3[2]

(141.0, 174.76070684814292, -33.76070684814292)

### Обучение градиентного бустинга из sklearn

<img src="https://drive.google.com/uc?id=1WCvGy1n3t_omeoE48-CC5GS9WsVBH1Yo" width=800>

<img src="https://drive.google.com/uc?id=1hvlxUhEaLmvABrKWgv4BDy0E-8noSFYN" width=800>

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor

boosting = GradientBoostingRegressor(
    learning_rate=1,
    n_estimators=100,
    random_state=10
)

In [ ]:
%%time
boosting.fit(X_train, y_train)

CPU times: user 1.32 s, sys: 3.88 ms, total: 1.33 s
Wall time: 2.49 s


GradientBoostingRegressor(learning_rate=1, random_state=10)

In [ ]:
boosting.score(X_train, y_train), boosting.score(X_test, y_test)

(0.9975984364178632, 0.8262436612002945)

#### lightgbm

<img src="https://drive.google.com/uc?id=1o6-ciHPtxodQNeUcCuyyVI2VY8rvxpPh">

*Exclusive Feature Bundling (EFB)*

Разряжённые данные значат, что многие признаки никогда ненулевые вместе. Это позволяет совместить ("bundle") эти фичи в одну без потери информации.<br>
Пример, у фичи А диапазон значений (0, 10) и у B - (0, 20). Добавляется смещение 10 к значениям фичи B, теперь их диапазон (10, 30). После, можно соединять A и B и использовать "feature bundle" с диапазоном (0, 30), чтобы заменить оригинальные A и B.

In [ ]:
from lightgbm import LGBMRegressor

lgbm = LGBMRegressor(
    random_state=10
)

In [ ]:
%%time
lgbm.fit(X_train, y_train)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001053 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 3094
[LightGBM] [Info] Number of data points in the train set: 1022, number of used features: 186
[LightGBM] [Info] Start training from score 181312.692759
CPU times: user 213 ms, sys: 3.53 ms, total: 217 ms
Wall time: 215 ms


LGBMRegressor(random_state=10)

In [ ]:
lgbm.score(X_train, y_train), lgbm.score(X_test, y_test)

(0.9761850066240383, 0.9000974446700329)

#### xgboost

<img src="https://drive.google.com/uc?id=1EURpT0NVlaSxQOgdI3p-sPbzrIEeF29w" width=550px>

<center><h4><i>Похожесть</i></h4></center><br>
$$Similarity = \frac{(\sum{Resideal_i})^2}{\sum{[Prev Proba_i * (1 - Prev Proba_i)]} + \lambda}$$

In [ ]:
0.25 / (0.25 + 1)

0.2

$Similarity_{10} =\frac{(-0.5 + 0.5 + 0.5 -0.5)^2}{[(0.5 * (1 - 0.5)) + (0.5 * (1 - 0.5)) + (0.5 * (1 - 0.5)) + (0.5 * (1 - 0.5))] + 0} = 0$<br>
$Similarity_{11} =\frac{(-0.5 + 0.5 + 0.5)^2}{[(0.5 * (1 - 0.5)) + (0.5 * (1 - 0.5)) + (0.5 * (1 - 0.5))] + 0} = 0.33$<br>
$Similarity_{21} =\frac{(-0.5)^2}{0.5 * (1 - 0.5) + 0} = 1$

<center><h4><i>Прирост</i></h4></center><br>
$$Gain = Left_{Similarity} + Right_{Similarity} - Root_{Similarity}$$

$Gain = 0.33 + 1 - 0 = 1.33$

<center><h4><i>Стрижка деревьев</i></h4></center><br>
$$Gain - \gamma \geqslant 0 \Rightarrow оставляем$$
$$Gain - \gamma < 0 \Rightarrow убираем$$

$\lambda$ - регуляризация<br>
  * default=1

$\gamma$ - минимальный прирост, чтобы не убирать разделение<br>
  * default=0
  - range: [0,∞]

$\eta$ - learning rate (скорость обучения)<br>
  - default=0.3
  - range: [0,1]

In [ ]:
from xgboost import XGBRegressor

xgb = XGBRegressor(
    random_state=10
)

In [ ]:
%%time
xgb.fit(X_train, y_train)

CPU times: user 2.68 s, sys: 7.82 ms, total: 2.69 s
Wall time: 2.23 s


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=10, ...)

In [ ]:
xgb.score(X_train, y_train), xgb.score(X_test, y_test)

(0.9998237781862409, 0.8986904707264346)

#### catboost

<img src="https://drive.google.com/uc?id=1fV7uyJxoP7cgo3lOp25cUuHKqm4cghlP">

In [ ]:
# 15 obj
# 1 tree 10 -> 5 obj error
# 2 tree 10 -> 5 obj error
# 3 tree 10 -> 5 obj error

- Предотвращение переобучения
    - Обучается несколько модель за одну итерацию и высчитываются остатки по наблюдениям на тех моделях, которые не видели его.
    - Высокий уровень рандома (перемешивание выборки и много случайности в начале построения дерева)
  
- Работа с категориальными признаками

<img src="https://drive.google.com/uc?id=1MYEseqxrQTqPbt81ZXSJt82YMF2WgW0y">

In [ ]:
!pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 MB 8.1 MB/s eta 0:00:00


In [ ]:
from catboost import CatBoostRegressor

cat = CatBoostRegressor(random_state=10)

In [ ]:
%%time
cat.fit(X_train, y_train)

Learning rate set to 0.041084
0:	learn: 75606.6914484	total: 64.6ms	remaining: 1m 4s
1:	learn: 73727.8595815	total: 79.4ms	remaining: 39.6s
2:	learn: 71982.0304482	total: 93.2ms	remaining: 31s
3:	learn: 70286.7504620	total: 102ms	remaining: 25.3s
4:	learn: 68590.7520022	total: 115ms	remaining: 22.8s
5:	learn: 67040.5236280	total: 127ms	remaining: 21s
6:	learn: 65468.7903371	total: 139ms	remaining: 19.7s
7:	learn: 64110.3021394	total: 152ms	remaining: 18.8s
8:	learn: 62749.7749189	total: 161ms	remaining: 17.8s
9:	learn: 61493.4426755	total: 169ms	remaining: 16.7s
10:	learn: 60138.6854091	total: 174ms	remaining: 15.7s
11:	learn: 58819.7898449	total: 182ms	remaining: 15s
12:	learn: 57613.0916028	total: 196ms	remaining: 14.9s
13:	learn: 56339.2498215	total: 205ms	remaining: 14.4s
14:	learn: 55310.6662027	total: 211ms	remaining: 13.8s
15:	learn: 54164.3909310	total: 219ms	remaining: 13.4s
16:	learn: 53152.4572944	total: 226ms	remaining: 13.1s
17:	learn: 52167.5664906	total: 232ms	remaining:

In [ ]:
cat.score(X_train, y_train), cat.score(X_test, y_test)

(0.9952603447277865, 0.9193565066045533)

[Гиперпараметры catboost](https://catboost.ai/docs/concepts/python-reference_parameters-list.html#python-reference_parameters-list)

In [ ]:
from google.colab import output
output.enable_custom_widget_manager()

In [ ]:
%%time
model_catb = CatBoostRegressor(silent=True, random_state=21,
                               eval_metric='R2',
                              #  cat_features=CAT_FEATURE_NAMES,
                               early_stopping_rounds=20,
                               use_best_model=True,
                               custom_metric=['RMSE', 'MAE']
                              )
model_catb.fit(X_train, y_train, plot=True, eval_set=(X_test, y_test))

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

CPU times: user 5.65 s, sys: 419 ms, total: 6.07 s
Wall time: 6.51 s


In [ ]:
model_catb.score(X_train, y_train), model_catb.score(X_test, y_test)

(0.9853142848487139, 0.9158610405617645)

In [ ]:
data = pd.read_csv('train.csv')
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(data.drop(columns=['SalePrice', 'Id']),
                                                    data['SalePrice'],
                                                    test_size=0.3,
                                                    random_state=42)
X_train.shape, X_test.shape

((1022, 79), (438, 79))

In [ ]:
categorials = X_train.select_dtypes('object').columns

X_train[categorials] = X_train[categorials].fillna('nan')
X_test[categorials] = X_test[categorials].fillna('nan')

In [ ]:
integer = X_train.select_dtypes(['int', 'float']).columns

X_train[integer] = X_train[integer].fillna(-999)

X_test[integer] = X_test[integer].fillna(-999)

In [ ]:
%%time
model_catb = CatBoostRegressor(silent=True, random_state=21,
                               eval_metric='R2',
                               cat_features=list(categorials),
                               early_stopping_rounds=20,
                               use_best_model=True,
                               custom_metric=['RMSE', 'MAE']
                              )
model_catb.fit(X_train, y_train, plot=True, eval_set=(X_test, y_test))

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

CPU times: user 30 s, sys: 2.83 s, total: 32.8 s
Wall time: 28.8 s


In [ ]:
model_catb.score(X_train, y_train), model_catb.score(X_test, y_test)

(0.9723473821472837, 0.9175949547329999)

# Дополнительные материалы:
1. Ансамблевые методы: бэггинг, бустинг и стекинг - https://neurohive.io/ru/osnovy-data-science/ansamblevye-metody-begging-busting-i-steking/
3. Бустинг - https://medium.com/mlreview/gradient-boosting-from-scratch-1e317ae4587d
6. CatBoost против Light GBM против XGBoost - https://machinelearningmastery.ru/catboost-vs-light-gbm-vs-xgboost-5f93620723db/
